openvdb_points/tools/AttributeGroup.h

```diff
@@ -0,0 +1,226 @@
+
+}; // class GroupAttributeArray
+
+
+inline GroupAttributeArray&
+GroupAttributeArray::cast(AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>()) {
+        OPENVDB_THROW(TypeError, "Invalid Attribute Type");
+    }
+    return static_cast<GroupAttributeArray&>(attributeArray);
+}
+
+
+inline const GroupAttributeArray&
+GroupAttributeArray::cast(const AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>()) {
+        OPENVDB_THROW(TypeError, "Invalid Attribute Type");
+    }
+    return static_cast<const GroupAttributeArray&>(attributeArray);
+}
+
+
+inline bool
+GroupAttributeArray::isGroup(const AttributeArray& attributeArray)
+{
+    if (!attributeArray.isType<GroupAttributeArray>())  return false;
+
+    return GroupAttributeArray::cast(attributeArray).isGroup();
+}
+
+
+////////////////////////////////////////
+
+
+class GroupHandle
+{
+public:
+    // Dummy class that distinguishes an offset from a bitmask on construction
+    struct BitMask { };
+
+    typedef std::pair<size_t, uint8_t> GroupIndex;
+
+    GroupHandle(const GroupAttributeArray& array,
+                const GroupType& offset)
+        : mArray(array)
+        , mBitMask(GroupType(1) << offset) { assert(mArray.isGroup()); }
+
+    GroupHandle(const GroupAttributeArray& array,
+                const GroupType& bitMask,
+                BitMask)
+        : mArray(array)
+        , mBitMask(bitMask) { assert(mArray.isGroup()); }
+
+    size_t size() const { return mArray.size(); }
+
+    bool get(Index n) const;
+
+private:
+    const GroupAttributeArray& mArray;
+    const GroupType mBitMask;
+}; // class GroupHandle
+
+
+////////////////////////////////////////
+
+
+class GroupWriteHandle
+{
+public:
+    typedef std::pair<size_t, uint8_t> GroupIndex;
+
+    GroupWriteHandle(   GroupAttributeArray& array,
+                        const GroupType& offset)
+        : mArray(array)
+        , mBitMask(GroupType(1) << offset) { assert(mArray.isGroup()); }
+
+    size_t size() const { return mArray.size(); }
+
+    bool get(Index n) const;
+
+    void set(Index n, bool on);
+
+private:
+    GroupAttributeArray& mArray;
+    const GroupType mBitMask;
+}; // class GroupWriteHandle
+
+
+////////////////////////////////////////
+
+
+/// Index filtering on group membership
+class GroupFilter
+{
+public:
+    GroupFilter(const GroupHandle& handle)
+        : mHandle(handle) { }
+
+    bool valid(const Index32 offset) const {
+        return mHandle.get(offset);
+    }
+
+private:
+    const GroupHandle mHandle;
+}; // class GroupFilter
+
+
+////////////////////////////////////////
+
+
+class GroupFilterFromLeaf
+{
+public:
+    typedef GroupFilter Filter;
+
+    GroupFilterFromLeaf(const Name& name)
+        : mName(name) { }
+
+    template <typename LeafT>
+    Filter fromLeaf(const LeafT& leaf) const {
+        return Filter(leaf.groupHandle(mName));
+    }
```

Comment:
You need to store "name" somewhere though in order to use it to extract the filter from the leaf.

Think about how you'd extend this to produce a filter that checked membership of two groups (instead of one). By putting this logic in a separate class that's used as a template parameter it makes this generic and extensible. Does that make sense?